gatsby-node.js

```diff
@@ -35,27 +42,93 @@ exports.createPages = async ({ actions, graphql, reporter }) => {
   const postOrPage = result.data.allMarkdownRemark.edges
     .filter(edge => edge.node.frontmatter.template !== "announcement");
 
+  // Render pages based on their frontmatter template, if available, or a
+  // default template at src/templates/default.js
   postOrPage.forEach(({ node }) => {
-    debugger;
     const templateName = node.frontmatter.template || `default`
     const template = path.resolve(path.join('src/templates/', `${templateName}.js`))
     createPage({
       path: node.fields.slug,
       component: template,
-      context: { id: node.id }, // additional data can be passed via context
+      // additional data can be passed via context
+      context: { id: node.id, locale: node.fields.locale },
     })
   })
 }
 
-exports.onCreateNode = ({ node, actions, getNode }) => {
+function localeFromFilename(filename) {
+  const potentialMatch = filename.match(/.*\.(\w+)\.\w+$/i)
+  return potentialMatch && potentialMatch[1] || null
+}
+
+function localePath(path, locale) {
+  let pathComponents = path.split('/').filter(s => s !== '')
+  let last = pathComponents.pop()
+
+  if (last === `index.${locale}`) {
+    last = ''
+  } else {
+    last = last.replace(new RegExp(`\\.${locale}$`, 'i'), '')
+  }
+
+  pathComponents.push(last)
+
+  return `/${locale}/${pathComponents.join('/')}`
+}
+
+exports.onCreateNode = async ({ graphql, node, actions, getNode }) => {
   const { createNodeField } = actions
 
+  // set a slug for all markdown nodes, and use any supported 639-1 language
+  // code prepended to the file extension to set a language-specific URL root
   if (node.internal.type === `MarkdownRemark`) {
-    const value = createFilePath({ node, getNode })
+    let locale = localeFromFilename(node.fileAbsolutePath)
+
+    if (!supportedLocales.includes(locale)) {
+      locale = defaultLocale
+    }
+
+    let slug = createFilePath({ node, getNode })
+
+    if (locale !== defaultLocale) {
```

Comment:
Flow is much nicer! 🎩 ✨ 

Also originally I only pointed out the `languagePath`/`localePath` thing because `languagePath` was never defined but `localePath` was. Whew you went the extra mile renaming everything to locale!! 💪